server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java

```diff
@@ -255,58 +255,70 @@ long findMaxStateId(final String prefix, Path... locations) throws IOException {
+                    try (DirectoryStream<Path> paths = Files.newDirectoryStream(stateDir)) {
+                        for (Path stateFile : paths) {
+                            final Matcher matcher = stateFilePattern.matcher(stateFile.getFileName().toString());
+                            if (matcher.matches()) {
+                                final long stateId = Long.parseLong(matcher.group(1));
+                                maxStateId = Math.max(maxStateId, stateId);
+                                PathAndStateId pav = new PathAndStateId(stateFile, stateId);
+                                logger.trace("found state file: {}", pav);
+                                files.add(pav);
+                            }
                         }
+                    } catch (NoSuchFileException | FileNotFoundException ex) {
+                        // no _state directory -- move on
                     }
-                } catch (NoSuchFileException | FileNotFoundException ex) {
-                    // no _state directory -- move on
                 }
             }
-        }
-        // NOTE: we might have multiple version of the latest state if there are multiple data dirs.. for this case
-        //       we iterate only over the ones with the max version.
-        long finalMaxStateId = maxStateId;
-        Collection<PathAndStateId> pathAndStateIds = files
+            // NOTE: we might have multiple version of the latest state if there are multiple data dirs.. for this case
+            //       we iterate only over the ones with the max version.
+            long finalMaxStateId = maxStateId;
+            Collection<PathAndStateId> pathAndStateIds = files
                 .stream()
                 .filter(pathAndStateId -> pathAndStateId.id == finalMaxStateId)
                 .collect(Collectors.toCollection(ArrayList::new));
 
-        final List<Throwable> exceptions = new ArrayList<>();
-        for (PathAndStateId pathAndStateId : pathAndStateIds) {
-            try {
-                T state = read(namedXContentRegistry, pathAndStateId.file);
-                logger.trace("state id [{}] read from [{}]", pathAndStateId.id, pathAndStateId.file.getFileName());
-                return state;
-            } catch (Exception e) {
-                exceptions.add(new IOException("failed to read " + pathAndStateId.toString(), e));
-                logger.debug(() -> new ParameterizedMessage(
+            final List<Throwable> exceptions = new ArrayList<>();
+            for (PathAndStateId pathAndStateId : pathAndStateIds) {
+                try {
+                    T state = read(namedXContentRegistry, pathAndStateId.file);
+                    logger.trace("state id [{}] read from [{}]", pathAndStateId.id, pathAndStateId.file.getFileName());
+                    return state;
+                } catch (NoSuchFileException | FileNotFoundException e) {
```

Comment:
This can only happen if a concurrent delete occurred I think. Just starting over with new `files` should work out eventually.